### ISSUE TITLE
Directories from glob are not marked as directories

### ISSUE THREAD

**Issue by UebelAndre** [bazelbuild/bazel#12954]
### Description of the problem / feature request:

Directories included from [glob](https://docs.bazel.build/versions/master/be/functions.html#glob) using `exclude_directories = 0` do not correctly return `True` for [File.is_directory](https://docs.bazel.build/versions/master/skylark/lib/File.html#is_directory).

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Running `bazel build ...` with the following files will reproduce this issue.

#### BUILD.bazel
```python
load("//:rule.bzl", "my_rule")

filegroup(
    name = "srcs",
    srcs = glob(
        ["**/**"],
        exclude_directories = 0,
    ),
)

my_rule(
    name = "my_rule",
    data = ":srcs",
)
```

#### rule.bzl
```python
def _rule_impl(ctx):
    for file in ctx.files.data:
        if file.is_source:
            print("{} is source".format(file))
        else:
            print("{} is directory".format(file))

my_rule = rule(
    implementation = _rule_impl,
    attrs = {
        "data": attr.label(
            allow_files = True,
            default = "//:srcs",
        )
    }
)
```

#### dir/file.txt
```python
# This is a file in a directory
```

#### WORKSPACE.bazel
```python
workspace(name = "is_directory_test")
```

### What operating system are you running Bazel on?

MacOS/Linux

### What's the output of `bazel info release`?

release 4.0.0

### Any other information, logs, or outputs that you want to share?

Note that `<source file dir> is source` from the output below is incorrect. It's a directory. `<source file dir> is directory` (or something to this effect, hitting the `else` condition in `rule.bzl`) should have been printed.
```command
bazel build ...
```
```output
INFO: Invocation ID: c54083cd-e13a-4092-91a5-90c4e4aef70c
DEBUG: /Users/user/Code/test/is_directory/rule.bzl:4:18: <source file BUILD.bazel> is source
DEBUG: /Users/user/Code/test/is_directory/rule.bzl:4:18: <source file WORKSPACE.bazel> is source
DEBUG: /Users/user/Code/test/is_directory/rule.bzl:4:18: <source file dir> is source
DEBUG: /Users/user/Code/test/is_directory/rule.bzl:4:18: <source file dir/file.txt> is source
DEBUG: /Users/user/Code/test/is_directory/rule.bzl:4:18: <source file rule.bzl> is source
INFO: Analyzed 2 targets (1 packages loaded, 7 targets configured).
INFO: Found 2 targets...
INFO: Elapsed time: 0.257s, Critical Path: 0.01s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
```

**Comment by aiuto** [comment]
I got bit by this today too.  It is not just from glob, it happens with explicit paths to directories as well.
Here a a reproduction.  https://github.com/aiuto/samples/tree/main/bazel_issue_12954

Is there any chance of bumping this up from P4?

**Comment by pauldraper** [comment]
Bazel just doesn't work with "source" directories.

That goes for is_directory, directory expansion in args, anything AFAIK.

If you want to use a directory, it has to come as the output of a rule.

**Comment by pauldraper** [comment]
> I got bit by this today too. It is not just from glob, it happens with explicit paths to directories as well.

Correct. `glob()` simply returns a list of strings, with which you can pass as files, strings, execute conditional logic, etc.

The correct title of this issue is "Source directories are not marked as is_directory"

I know that source directories aren't really supported in Bazel. I was using them for efficient processing of hundreds of npm dependencies. There isn't really a great alternative.

**Comment by pauldraper** [comment]
Comment

**Comment by cameron-martin** [comment]
A related issue, with possibly the same root cause: `DirectoryExpander`, as used in the `Args` object, does not expand source directories.

**Comment by pauldraper** [comment]
@cameron-martin , yes. This is what I meant when I said

> This [affects] directory expansion in args as well.
